### ISSUE TITLE
For some actions, originalName field doesn't have original name as in P4 program. Is this the correct behavior?

### ISSUE THREAD

**Issue by usha1830** [p4lang/p4c#3171]
Isn't the originalName field of ID supposed to remain unchanged throughout the compilation flow? 
It gets modified by some frontend pass.

The below test has some actions where the originalName itself gets modified.
https://github.com/p4lang/p4c/blob/main/testdata/p4_16_samples/pna-example-tunnel.p4

**Comment by mbudiu-vmw** [comment]
The originalName has no reason to change.
What is the symptom you see that is incorrect?

**Comment by usha1830** [comment]
To confirm that the originalName is changed, I had simply added new visitors in some of the frontend passes for P4Action and ActionListElement printing the originalName.


To give some context, there is a mismatch between the action name in the action declaration and the action name in the action list. It caused an issue while optimizing away the unused actions in DPDK backend(some of the actions had the originalName modified and was being away because the name wasn't found in any table's ActionList)

I had tried to fix this issue by using the internal name as obtained using IR::ID's name instead of using toString() in this [PR](https://github.com/p4lang/p4c/pull/3109).
toString() uses the originalName

`cstring toString() const override { return originalName.isNullOrEmpty() ? name : originalName; }`

But this is causing an issue with the control plane software which needs the action names to match across various outputs (for DPDK, we have 4 output files -.spec, bfrt.json, context json and P4Info coming from the frontend). The runtime files use the controlPlaneName() which again boils down to using the toString() if name annotation is not present.

If originalName is not supposed to change throughout, I was thinking we can use revert back to using toString() in dpdk .spec file generation as used in the runtime files generation.

Do you have any other suggestions?

I will look into why the originalName is changed for some actions. Probably this is done in an attempt make the name unique.

**Comment by mbudiu-vmw** [comment]
For that example what is a name that you find changed?
Note that the compiler can duplicate actions that are used in multiple tables, and each will have a new name, but I expect that the originalName is still unchanged.

**Comment by usha1830** [comment]
For this example:
https://github.com/p4lang/p4c/blob/main/testdata/p4_16_samples/pna-example-tunnel.p4

The actions that have modified originalName are
decap_outer_ipv4 -> tunnel_decap_decap_outer_ipv4
set_tunnel -> tunnel_encap_set_tunnel

Pasting the code from the example here for quick reference
```
* Tunnel Decap */
control tunnel_decap(inout headers_t hdr,
                     inout local_metadata_t local_metadata)
{
  action decap_outer_ipv4(tunnel_id_t tunnel_id) {
    local_metadata.tunnel.id = tunnel_id;
  }

  table ipv4_tunnel_term_table {
    key = {
      hdr.outer_ipv4.src_addr : exact @name("ipv4_src");
      hdr.outer_ipv4.dst_addr : exact @name("ipv4_dst");
      local_metadata.tunnel.tun_type : exact;
    }
    actions = {
      decap_outer_ipv4;
      @defaultonly NoAction;
    }
    const default_action = NoAction;
  }

  apply {
    ipv4_tunnel_term_table.apply();
  }
}  // control tunnel_decap

/* Tunnel Encap */
control tunnel_encap(inout headers_t hdr,
               inout local_metadata_t local_metadata,
               in pna_main_input_metadata_t istd)
{
  action set_tunnel(ipv4_addr_t dst_addr) {
    local_metadata.outer_ipv4_dst = dst_addr;
  }

  table set_tunnel_encap {
    key = {
      istd.input_port: exact;
    }
    actions = {
      set_tunnel;
      @defaultonly NoAction;
    }
    const default_action = NoAction;
    size = 256;
  }

  apply {
    set_tunnel_encap.apply();
  }
}
```

**Comment by mbudiu-vmw** [comment]
You should think of these as really new actions synthesized from the old ones.
Please note that they both have a `@name` annotation which preserves information about the original name:
```
    @name("tunnel_decap.decap_outer_ipv4") action tunnel_decap_decap_outer_ipv4(@name("tunnel_id") tunnel_id_t tunnel_id_0) {
        local_metadata.tunnel.id = tunnel_id_0;
    }
```
I would have to think a bit whether we could actually preserve the originalName too... I don't remember very well how it is really used. But perhaps the annotation can give you all the information you need?

**Comment by jafingerhut** [comment]
I have no idea if this issue is relevant to what you are working on, but thought I would link to it from here, in case it is: https://github.com/p4lang/p4c/issues/2755

**Comment by usha1830** [comment]
> You should think of these as really new actions synthesized from the old ones. Please note that they both have a `@name` annotation which preserves information about the original name:
> 
> ```
>     @name("tunnel_decap.decap_outer_ipv4") action tunnel_decap_decap_outer_ipv4(@name("tunnel_id") tunnel_id_t tunnel_id_0) {
>         local_metadata.tunnel.id = tunnel_id_0;
>     }
> ```
> 
> I would have to think a bit whether we could actually preserve the originalName too... I don't remember very well how it is really used. But perhaps the annotation can give you all the information you need?

Yes. I think the @name annotation can be used. Probably, if @name is present, use that, otherwise use originalName.
Thanks!
